internal/nvmeof/nvmeof.go

```diff
@@ -0,0 +1,417 @@
+	log.DebugLog(ctx, "Adding host %s to subsystem %s on gateway %s",
+		hostNQN, subsystemNQN, gw.config)
+
+	req := &AddHostReq{
+		SubsystemNqn: subsystemNQN,
+		HostNqn:      hostNQN,
+		// Psk: nil,          // No pre-shared key
+		// DhchapKey: nil,    // No DH-CHAP authentication
+		// PskEncrypted: nil, // No PSK encryption
+		// KeyEncrypted: nil, // No key encryption
+	}
+
+	resp, err := gw.client.AddHost(ctx, req)
+	if err != nil {
+		return fmt.Errorf("failed to add host %s to subsystem %s: %w", hostNQN, subsystemNQN, err)
+	}
+	if resp.GetStatus() != 0 {
+		return fmt.Errorf("gateway AddHost returned error (status=%d): %s", resp.GetStatus(), resp.GetErrorMessage())
+	}
+	log.DebugLog(ctx, "Host added successfully: %s to subsystem %s", hostNQN, subsystemNQN)
+
+	return nil
+}
+
+func (gw *GatewayRpcClient) CreateListener(ctx context.Context, subsystemNQN string, listenerInfo ListenerDetails,
+) error {
+	log.DebugLog(ctx, "Adding listener %s to subsystem %s", listenerInfo.Address, subsystemNQN)
+	adrfam := AddressFamily_ipv4
+	req := &CreateListenerReq{
+		Nqn:      subsystemNQN,
+		HostName: listenerInfo.Hostname,
+		Traddr:   listenerInfo.Address,
+		Trsvcid:  &listenerInfo.Port,
+		Adrfam:   &adrfam, // Assuming IPv4, can be configurable // TODO - make it configurable
+		// Secure  // false, // Assuming no security for now   // TODO - make it configurable?
+		// VerifyHostName // false, // Assuming no hostname verification for now // TODO - make it configurable
+	}
+
+	resp, err := gw.client.CreateListener(ctx, req)
+	if err != nil {
+		return fmt.Errorf("failed to add listener %s to subsystem %s: %w", listenerInfo.Address, subsystemNQN, err)
+	}
+	if resp.GetStatus() != 0 {
+		return fmt.Errorf("gateway AddListener returned error (status=%d): %s", resp.GetStatus(), resp.GetErrorMessage())
+	}
+
+	log.DebugLog(ctx, "Listener added successfully: %s to subsystem %s", listenerInfo.Address, subsystemNQN)
+
+	return nil
+}
+
+// DeleteListener removes a listener from a subsystem.
+func (gw *GatewayRpcClient) DeleteListener(ctx context.Context, subsystemNQN string, listenerInfo ListenerDetails,
+) error {
+	log.DebugLog(ctx, "Deleting listener %s from subsystem %s", listenerInfo.Address, subsystemNQN)
+	adrfam := AddressFamily_ipv4
+	// Add this because without this, the gw return
+	// "Can't verify there are no active connections for this address" instead of "Listener not found"
+	force := true
+	req := &DeleteListenerReq{
+		Nqn:      subsystemNQN,
+		HostName: listenerInfo.Hostname,
+		Traddr:   listenerInfo.Address,
+		Trsvcid:  &listenerInfo.Port,
+		Adrfam:   &adrfam, // Assuming IPv4, can be configurable // TODO - make it configurable
+		Force:    &force,
+	}
+
+	resp, err := gw.client.DeleteListener(ctx, req)
+	if err != nil {
+		return fmt.Errorf("failed to delete listener %s from subsystem %s: %w", listenerInfo.Address, subsystemNQN, err)
+	}
+	if resp.GetStatus() == 0 {
```

Comment:
it would be nice to return error early and avoid multiple if if we check `resp.GetStatus() != 0 && resp.GetStatus() != int32(syscall.ENOENT)` same is required in other Delete calls as we..

Reply:
i assume this is for logging, IMO we dont need to log if it is success or NotFound because its same for CSI where the deletion is successful. if really required we can log the status code at the end which helps for debugging. its not worth to add a if statement to log different error messages.